src/ethereum_test/types.py

```diff
@@ -318,14 +394,137 @@ def to_geth_dict(self) -> Mapping[str, Any]:
         return header
 
 
+@dataclass(kw_only=True)
+class Header:
+    """
+    Header type used to describe block header properties in test specs.
+    """
+
+    coinbase: Optional[str] = None
+    parent_hash: Optional[str] = None
+    ommers_hash: Optional[str] = None
+    state_root: Optional[str] = None
+    transactions_root: Optional[str] = None
+    receipt_root: Optional[str] = None
+    bloom: Optional[str] = None
+    difficulty: Optional[int] = None
+    number: Optional[int] = None
+    gas_limit: Optional[int] = None
+    gas_used: Optional[int] = None
+    timestamp: Optional[int] = None
+    extra_data: Optional[str] = None
+    mix_digest: Optional[str] = None
+    nonce: Optional[str] = None
+    base_fee: Optional[int] = None
+    hash: Optional[str] = None
+
+
+@dataclass(kw_only=True)
+class Block(Header):
+    """
+    Block type used to describe block properties in test specs
+    """
+
+    rlp: Optional[str] = None
+    """
+    If set, blockchain test will skip generating the block using
+    `evm_block_builder`, and will pass this value directly to the Fixture.
+
+    Only meant to be used to simulate blocks with bad formats, and therefore
+    requires the block to produce an exception.
+    """
+    rlp_modifier: Optional[Header] = None
+    """
+    An RLP modifying header which values would be used to override the ones
+    returned by the  `evm_transition_tool`.
+    """
+    exception: Optional[str] = None
+    """
+    If set, the block is expected to be rejected by the client.
+    """
+    txs: Optional[List[Transaction]] = None
+    """
+    List of transactions included in the block.
+    """
+    uncles: Optional[List[Header]] = None
+    """
+    List of uncle headers included in the block.
+    """
+
+    def set_environment(self, env: Environment) -> Environment:
+        """
+        Creates a copy of the environment with the characteristics of this
+        specific block.
+        """
+        new_env = copy(env)
+
+        """
+        Values that need to be set in the environment and are `None` for
+        this block need to be set to their defaults.
+        """
+        environment_default = Environment()
+        new_env.difficulty = self.difficulty
+        new_env.coinbase = (
+            self.coinbase
+            if self.coinbase is not None
+            else environment_default.coinbase
+        )
+        new_env.gas_limit = (
+            self.gas_limit
+            if self.gas_limit is not None
+            else environment_default.gas_limit
+        )
+        new_env.base_fee = self.base_fee
+
+        """
+        These values are required, but they depend on the previous environment,
+        so they can be calculated here.
+        """
+        if self.number is not None:
+            new_env.number = self.number
+        else:
+            # calculate the next block number for the environment
+            if len(new_env.block_hashes) == 0:
+                new_env.number = 1
```

Comment:
This assumes the genesis block isn't in the env block hashes?

Reply:
Fixed this error. Thanks!